### ISSUE TITLE
re-configuration loop.

### ISSUE THREAD

**Issue by LelandSindt** [vfarcic/docker-flow-proxy#453]
**Description**

When multiple services are being started/stopped in rapid succession the HAProxy config verification can fail verifying previously configured services causing Proxy and Swarm-Listener to fall into an endless re-configuration loop. During this time no new/existing service state changes are seen by Swarm-Listener

**Steps to reproduce the issue:**
1. clone https://github.com/LelandSindt/docker-flow-chaos-monkey.git
`git clone https://github.com/LelandSindt/docker-flow-chaos-monkey.git`
2. start the docker-flow proxy stack... 
`cd docker-flow-chaos-monkey/ && ./start_docker_flow.sh `
3. start the failure reproduction script... and follow the on-screen instructions... 
`cd 0001/ && ./001_fail.sh`

example output... 

```
leland@instance-1:~/src$ cd docker-flow-chaos-monkey/ && ./start_docker_flow.sh 
e6cwpwo0qgty        proxy                   overlay             swarm
Creating service proxy_swarm-listener
Creating service proxy_proxy
leland@instance-1:~/src/docker-flow-chaos-monkey$ cd 0001/ && ./001_fail.sh
start the first service quickly....
Starting one. Start Delay: 0. Health Delay: 1. Replicas: 1
kth9bpyzv3jycxdxb1hmphgjb
swarm-listener will see "one", notify proxy and proxy will complete the configuration. 
waiting for "one" to come online......
start second service slowly...
Starting two. Start Delay: 30. Health Delay: 30. Replicas: 1
882zz5drx2iebnqioqz39wxt9
swarm-listener will see "two" and notify proxy, proxy will fail to resolve "two" because it has not yet completed its health check and is not available to the swarm
waiting for "two" to come online...... timed out. (5)
remove the first service 
one
swarm listener is still waiting on/retrying the config for "two". Eventually "two" will come online but with "one" gone, proxy will fail to verify the congig... swarm-listener will contine to try to configure "two" ?forever?.... and will not see any other services come online/go offline  

the script will now wait forever for service "two" to come online. ctrl-c will cause the script to clean everything up... 
waiting for "two" to come online.....................^C** Trapped CTRL-C
bring "one" back online... 
Starting one. Start Delay: 0. Health Delay: 1. Replicas: 1
z5z8jts4wxo7uyxtjvmnl4x3g
waiting for "one" to come online. 
waiting for "two" to come online.
remove "one" and "two" 
one
two
```

**Describe the results you received:**
swarm-listener and proxy end up in an endless re-configuration loop while service "one" is offline as swarm-listener and proxy try to bring service "two" online.

**Describe the results you expected:**
I expected swarm-listener and proxy to re-configure gracefully... ;)

**Note**
Near as I can tell... Swarm-Listener expects to get a quick response from Proxy. However, with services that are slow to reach a running status, the Proxy config verification will fail. If during this time another configured service is taken offline it will cause the config verification to permanently fail. 

Would it be possible/advisable to have Swarm-Listener wait until a new service is running/healthy before notifying Proxy?

**Comment by thomasjpfan** [comment]
I'm currently looking into this.

**Comment by thomasjpfan** [comment]
I think `swarm-listener` needs to be updated as follows:

1. Wait for services to become healthy by default.
2. Remove service notifications need to be sent out right away. (No waiting required)
3. If a service is waiting to be healthy and it is removed, no notification is sent out.

**Comment by thomasjpfan** [comment]
This feature is implemented in `docker-flow-swarm-listener` with release tag `18.04.04-2` or later.

I just tested it out with your scripts and `proxy` is configured correctly after waiting for service "two" to become healthy.

**Comment by LelandSindt** [comment]
@thomasjpfan  --- thank you!

I will test this case tonight. I will also throw the chaos.sh script at it.

**Comment by LelandSindt** [comment]
I have run `while true; do ./001_fail.sh && docker service rm two; done` for hours without issue :+1: 

However, I threw in a couple of instances of chaos.sh and was once again able to wedge proxy/swarm-listener. 

I will review more as time allows and provide details/reproduction examples...

**Comment by LelandSindt** [comment]
The initial issue has clearly been addressed.

`while true; do ./001_fail.sh && docker service rm two; done` --or--- a single instance of `./chaos.sh` (found in https://github.com/LelandSindt/docker-flow-chaos-monkey.git) run without issue against `18.04.04-4`

I was able to wedge `18.04.04-4` once, using a couple instances of `chaos.sh` but I will open an new issue if/when I can reproduce it reliably.

**Comment by LelandSindt** [comment]
`while true; do ./001_fail.sh && docker service rm two; done` is working without issue. 

I will through `chaos.sh` at it overnight. note: I have found that`chaos.sh` /can/ (depending on how many monkeys it's configure to use... ) overwhelm docker swarm at which point you can't expect `swarm-listener` to perform.

**Comment by LelandSindt** [comment]
I forgot to kick off `chaos.sh` last night so `while true; do ./001_fail.sh && docker service rm two; done` ran all night, I didn't review all the logs, but there were no failures that wedged DFP or DFSL... 

I ran `chaos.sh` with 50 monkeys... and got DFP and DFSL into a re-configuration loop within ~10 minutes.
